src/core/qgsgrouplayerrenderer.cpp

```diff
@@ -0,0 +1,149 @@
+      continue;
+    }
+
+    mChildRenderers.emplace_back( childLayer->createMapRenderer( context ) );
+    mRendererCompositionModes.emplace_back( childLayer->blendMode() );
+    mRendererOpacity.emplace_back( childLayer->type() != QgsMapLayerType::RasterLayer ? childLayer->opacity() : 1.0 );
+    mTransforms.emplace_back( layerToDestTransform );
+  }
+
+  mPaintEffect.reset( layer->paintEffect() ? layer->paintEffect()->clone() : nullptr );
+}
+
+QgsGroupLayerRenderer::~QgsGroupLayerRenderer() = default;
+
+QgsFeedback *QgsGroupLayerRenderer::feedback() const
+{
+  return mFeedback.get();
+}
+
+bool QgsGroupLayerRenderer::render()
+{
+  QgsRenderContext &context = *renderContext();
+
+  context.painter()->save();
+  if ( mPaintEffect )
+  {
+    mPaintEffect->begin( context );
+  }
+
+  const QgsCoordinateReferenceSystem destinationCrs = context.coordinateTransform().destinationCrs();
+  bool canceled = false;
+  int i = 0;
+  for ( const std::unique_ptr< QgsMapLayerRenderer > &renderer : std::as_const( mChildRenderers ) )
+  {
+    if ( mFeedback->isCanceled() )
+    {
+      canceled = true;
+      break;
+    }
+
+    context.setCoordinateTransform( mTransforms[i] );
+
+    // don't need to catch exceptions here -- it would have already been caught in the QgsGroupLayerRenderer constructor!
+    const QgsRectangle extentInChildLayerCrs = mTransforms[i].transformBoundingBox( context.mapExtent(), Qgis::TransformDirection::Reverse );
+    context.setExtent( extentInChildLayerCrs );
+
+    QImage image;
+    QPainter *prevPainter = context.painter();
+    std::unique_ptr< QPainter > imagePainter;
+    if ( renderer->forceRasterRender() )
+    {
+      image = QImage( context.deviceOutputSize(), context.imageFormat() );
+      image.setDevicePixelRatio( static_cast<qreal>( context.devicePixelRatio() ) );
+      image.fill( 0 );
+      imagePainter = std::make_unique< QPainter >( &image );
+
+      context.setPainterFlagsUsingContext( imagePainter.get() );
+      context.setPainter( imagePainter.get() );
+    }
+    renderer->render();
+
+    if ( imagePainter )
+    {
+      imagePainter->end();
+      context.setPainter( prevPainter );
+      if ( context.useAdvancedEffects() )
+        context.painter()->setCompositionMode( mRendererCompositionModes[i] );
+
+      context.painter()->setOpacity( mRendererOpacity[i] );
+      context.painter()->drawImage( 0, 0, image );
+      context.painter()->setOpacity( 1.0 );
+      context.painter()->setCompositionMode( QPainter::CompositionMode_SourceOver );
+    }
+    i++;
+  }
+
+  if ( mPaintEffect )
+  {
+    mPaintEffect->end( context );
+  }
+
+  context.painter()->restore();
+
+  return !canceled;
+}
+
+bool QgsGroupLayerRenderer::forceRasterRender() const
+{
+  if ( !renderContext()->testFlag( Qgis::RenderContextFlag::UseAdvancedEffects ) )
+    return false;
+
+  if ( !qgsDoubleNear( mLayerOpacity, 1.0 ) )
```

Comment:
@nyalldawson , that just made me realize we'll be able to control grouped layers' opacity; that alone is worth being excited about :wink: